// @sealights/server-mock/dist/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
var jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
var uuid_1 = require("uuid");
var bogusPrivateKey = "-----BEGIN RSA PRIVATE KEY-----\nMIIEowIBAAKCAQEAjx1swza766cvNDglW1Y3sfBGfnSUQ6LKP1+B1ERCulOdhB78\n3KoP1nkZ93LfiJrLamNXH0irS5rhA1KUHTFR6RZq9Bp8SrGdFuraZGMHRNrN5tRY\natbGIqScuBk6P7x8YGckpxPxKloq3c2SixxLAbsGjVocd6vIEfae5g7GV7gPWgET\nxNVVyObwVi7UjMv9acnWAlEDSTHoEZDsphtagUmieIb/bf8uS2vdwYXAA9l3wg/J\nj2uebDcoNtI5YnO5Zc20+dG4cEH+vZQEyXoTRXEu5pk9iCvWi6Y/vPSv6F8pQCJ0\ny/+C0CEMJk7GcHzBCuXkhrekA4gI8B+neH7VWQIDAQABAoIBAH570gEh4anew0BS\nkumLZIPVyGfAnry8JiUmQrcYcST9+0umm3JmtsKlmwZwNxEjRb5L99bFQeb7Bf0K\nwa6fRTZMws/kjI7zNNPpUoTAZM881hc4+r/J+6jodT3bxfhInGATKHiljlIRCKup\nIOfjZIMuCRdqWZiGcLc3kK3gxvkEdOgQGZkdcKTFyR6UiMF19tTJ+EmRgZdCvnXe\nxkWfZoOdW3FymV0AOvEbVrEWrmpw30TshE8afCYDOPKMpCCHgPyMEaWVZJGTj/5g\nX79ktQPpJ7DnPaxzdOWJ+/t51SFYBvcQG1SSij9/TneSPKPzAUidTH6dBsqsaVED\nQiio6TUCgYEA2seYTzY9CvrdbfC+oJ2VNqhsc6gTl+RPinGvA8snMlPQAgFBA/v8\ngiB/aG9/P9NhvhUBnTUG8fBkOk4vhgKVkiEx6Rhc+Q8tBOQqRLAKuzzsqzfTZBi7\noAIqIB5ppJJG2tWpXK16PtCljIfpJtcUbUU5OT8ENY1AYZuBMrWPB/cCgYEAp3Zw\nzfT+MkzqMgm6jAw/u+tGOnvAeaB9BUhz/hzisrUwwdGVTjDA8hwW65oAGmDeJr1+\nXD90lX+bDMBOe4nJdAAKEwYGgOV+mkIAffL0O4t6+UjrJmhf+R1rw3Nu38c8rza+\nQza8v5JFUzRfEoSa44zlHlHCEpIuaF9zCz9k2S8CgYAxCgBH8R5psaWkmQwBsJDb\nuZDls80AhPqXS7gxH35ft42uhV1vaTJGcMZeQG/DfLwyAT9lX5d476tnrlLUtQuV\nv78SQpXEdfRfXk1jLBqbsaJTFqZZj3mu+g07OdXwP6GGN9JIpdH69oXFJEnMnPVF\nMPkMreVho5HEzzh8h1nSLQKBgQCZdw5bgEgvmpMI9idxzovdz/fVArHaBHxuCf7a\ngmgyM6ASDtpQBJfdiGZs1P2BLwKG6CNTsIdLtdxGrDmANz5Zwut4kY4GM8p43UGM\nKskthF5W8z8kPISvTAoE3NVf2DtBqt5jxFbnDMnhPV5fHwy+udECnxRGZpaG6eoJ\nBbiK2wKBgCXqMhywzo7imeN5LVqW2crVk9wPVq+jTnlS/IzLP7eR2cKreFVB7Ysj\nlCY7LXrnJHPJjlJR3LETXUCuvhOt+e3ZrUe/ZiouEWM8wVa83wrA949N9uMdUmfO\nZ5yaKzuJrITO10ohcmFLd/tKDUtv7P3XbC2xTNpVJiexdtQVcCtA\n-----END RSA PRIVATE KEY-----";
function getWebToken() {
    var port = parseInt(process.env.SERVER_PORT || '9900');
    var sl_server = 'localhost';
    if (require('os').type().toLowerCase().indexOf('win') === -1) {
        sl_server = require('os').hostname();
    }
    console.log("server: \"" + sl_server + "\"");
    var token = jsonwebtoken_1.default.sign({
        subject: 'SeaLights@agent',
        'x-sl-role': 'agent',
        'x-sl-server': "http://" + sl_server + ":" + port + "/api"
    }, bogusPrivateKey, { algorithm: 'RS512' });
    return token;
}
function sleep(duration) {
    return new Promise(function (resolve, reject) {
        setTimeout(resolve, duration);
    });
}
function createNewBuildSession(customerId, appName, branchName, buildName, additionalParams) {
    return {
        buildSessionId: uuid_1.v4().replace(/\-/g, ""),
        customerId: customerId,
        appName: appName,
        branchName: branchName,
        buildName: buildName,
        additionalParams: additionalParams,
    };
}
function sendResponse(res, status, type, body) {
    res.status(status).type(type).send(body);
    res.end();
}
var Utils = { getWebToken: getWebToken, sleep: sleep, createNewBuildSession: createNewBuildSession, sendResponse: sendResponse };
exports.Utils = Utils;
